### PR TITLE
Restore opened files between application restarts

### PR DESCRIPTION
**Pull request by dail8859**
This is a first attempt at saving/loading a user's session between application restarts.

This feature is **not** enabled by default. This needs enabled through the preferences dialog.

~~This feature does **not** currently save temporary files (that will come later). This base functionality is needed before temporary files can be properly saved/loaded.~~

TODO:
- [x] Make sure new files don't get named the same as existing files. E.g. since `New 1` can be loaded from the session, `File > New` would then also create `New 1`
- [x] Allow to files with the same name to be stored in the session. `C:\dir1\test.txt` and `C:\dir2\test.txt` both get saved as `text.txt` and will overwite each other.
- [x] Remove commented out code

### CODE REVIEW

File: src/NotepadNext/SessionManager.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Notepad Next.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+
+#ifndef SESSIONMANAGER_H
```
Comment (elcuco) [inline src/NotepadNext/SessionManager.h:20]: You should start using `#pragma once` it does the same, and is supported on GCC since forever. Do you want me to handle this in another PR?

Comment (dail8859) [reply src/NotepadNext/SessionManager.h:20]: > is supported on GCC since forever

Interesting. Guess I always thought of `#pragma once` as a MSVC++ specific feature (I first found it many years ago!), but as you mentioned it appears it has been universally supported for a while now.

So yes I agree `#pragma once` is desirable. You are certainly welcome to submit the changes. Only need to worry about code under `.src/NotepadNext`

**Comment by mikelupe** [comment]
Very nice, will be my daily driver - thanks for your efforts.

One situation ("bug"): 
0. Fresh start, no sessions apart "New 1"
1. "New 1" > type some text > "Save as.." test.txt
2. Double click to create a new tab > "New 1" > type some text > close NotepadNext w/o saving
3. Relaunch NotepadNext ; State: one tab "test.txt" (red with content), one tab "New 1" (red with content)
4. Again double click to create a new tab > "New 1" (blue empty)
5. On close and reopen, only the last "New 1" is kept, the original "New 1" (red) is ~overwritten

Basically: As long as a "New X" tab hasn't been saved, the next increment of new still is "New X", so no increment :)
I hope it's reproducable

kind regards

**Comment by dail8859** [comment]
@mikelupe Thanks for testing it out!

> 1. "New 1" > type some text > "Save as.." test.txt
> 2. Double click to create a new tab > "New 1" > type some text > close NotepadNext w/o saving

Just to make sure, in step 2 wouldn't it be "New 2" since the "New 1" document was already saved?

---

But yes this is one of the bugs, that you can't have two files with the same name currently. A couple points to this:

- If `New 2` (or any other temp file for that matter) is loaded, then a new tab (e.g. File > New) should never be named that, so some logic would need added to keep incrementing the number in the name until one is found that doesn't exist.
- The same thing will happen if you have `C:\dir1\test.txt` and `C:\dir2\test.txt` since they both tried to get saved in the session folder both as `test.txt`. I've got an idea to fix this.

**Comment by mikelupe** [comment]
> Just to make sure, in step 2 wouldn't it be "New 2" since the "New 1" document was already saved?

There would be no more "New 1" as it had been saved as "test.txt" in step 1 

Thanks again for your effort, this will be epic :)

**Comment by mikelupe** [comment]
I checked your latest commit:

I had to remove the rm ~/.config/NotepadNext/NotepadNext.ini , otherwise I got a segfault :) 

Unfortunately now it generates a "New 2" but then later on again another "New 2". Almost there

**Comment by dail8859** [comment]
> otherwise I got a segfault :)

Yeah that makes sense, but would indicate some better error handling is needed incase the settings file gets modified or corrupted.

> Unfortunately now it generates a "New 2" but then later on again another "New 2". Almost there

Yep the root issue was fixed so that even if you have two `New 2` temp files they won't overwrite each other. Still need to fix the file name incrementing to prevent this.

Just curious since you are on a non-Windows machine, where exactly does it save the session files too? I'm assuming something like `~/.config/NotepadNext/NotepadNext/session/`? I've only tested it on Windows.

**Comment by mikelupe** [comment]
I see, nice.

Re location - I had to find it first :) It saves it to "~/.local/share/NotepadNext/NotepadNext/session/" . so $HOME/.local/...

**Comment by dail8859** [comment]
> It saves it to "~/.local/share/NotepadNext/NotepadNext/session/" . so $HOME/.local/...

Hmmm, ok. The Qt documentation for standard paths is a bit fuzzy, so it saved to a semi-sane location on Windows but didn't know what that would look like on other systems.

**Comment by mikelupe** [comment]
I think that's quite legit. ~/.config/NotepadNext/session/ would be ok as well. Eventually a config setting with the recent defaults could make sense.

**Comment by mikelupe** [comment]
Nice

File: src/NotepadNext/EditorManager.h [hunk 2]
```diff
@@ -33,18 +33,19 @@ class EditorManager : public QObject
 public:
     explicit EditorManager(QObject *parent = nullptr);
 
-    ScintillaNext *createEmptyEditor(const QString &name);
+    ScintillaNext *createEditor(const QString &name);
     ScintillaNext *createEditorFromFile(const QString &filePath, bool tryToCreate=false);
-    ScintillaNext *cloneEditor(ScintillaNext *editor);
+    //ScintillaNext *cloneEditor(ScintillaNext *editor);
```
Comment (elcuco) [inline src/NotepadNext/EditorManager.h:38]: you "should" clean this up

Comment (dail8859) [reply src/NotepadNext/EditorManager.h:38]: Agreed

Comment (dail8859) [reply src/NotepadNext/EditorManager.h]: Fixed

**Comment by mikelupe** [comment]
Thanks for your speedy work, and for the mention :) The latest build is becoming my driver. 

There is one thing though, double clicking generates incremented "New X", but if closed w/o saving, on a new creation it still increments - so create "New 2" > close "New 2", create a new one gives "New 3" > close w/o saving > create a new one gives "New 4" etc. 

Would it be easy to check on this, even if it's not really a show-stopper? It would make it even more consistent.

**Comment by dail8859** [comment]
> There is one thing though, double clicking generates incremented "New X", but if closed w/o saving, on a new creation it still increments - so create "New 2" > close "New 2", create a new one gives "New 3" > close w/o saving > create a new one gives "New 4" etc.

This is actually a preexisting bug (not sure if this PR made it worse or not) but even in the official v0.5.6 release I can recreate it by:

1. Start a fresh instance, which just contains `New 1`
2. `File > New` creates `New 2`
3. `File > Close All` closes files and creates `New 3`
4. Repeat step 3, creates `New 4` etc.

This was one of the bugs I ran across and thought I introduced it with these changes but there is at least some not-desirable behavior that already exists. Appreciate you testing things and providing feedback!

**Comment by mikelupe** [comment]
Let's be frank, It's not really a bug, maybe just some polishing.

Something else, would a dark mode be a lot of work? But this would extend it over original Notepad++'s features.

Thanks again for your work

**Comment by dail8859** [comment]
> Let's be frank, It's not really a bug, maybe just some polishing.

Agreed

> Something else, would a dark mode be a lot of work?

Yes and no. Getting something basic is easy, to really polish it would probably take a bit of time. There may be existing 3rd party dark theme css files that would work well but I haven't tested them. You can load custom css files: see https://github.com/dail8859/NotepadNext/issues/171#issuecomment-1186036612

Not something I'll be working on myself any time soon.

There are other non-Windows systems that have Qt theming built-in so it carries over to Notepad Next: https://github.com/dail8859/NotepadNext/issues/109#issuecomment-1105907443

**Comment by mikelupe** [comment]
Thanks for the hint.

As it's mainly about the editor background that would need to be dark, I'll be waiting. There really is no hurry.
